### PR TITLE
podman cp: support copying on tmpfs mounts

### PR DESCRIPTION
**Pull request by vrothberg**
Traditionally, the path resolution for containers has been resolved on
the *host*; relative to the container's mount point or relative to
specified bind mounts or volumes.

While this works nicely for non-running containers, it poses a problem
for running ones.  In that case, certain kinds of mounts (e.g., tmpfs)
will not resolve correctly.  A tmpfs is held in memory and hence cannot
be resolved relatively to the container's mount point.  A copy operation
will succeed but the data will not show up inside the container.

To support these kinds of mounts, we need to join the *running*
container's mount namespace (and PID namespace) when copying.

Note that this change implies moving the copy and stat logic into
`libpod` since we need to keep the container locked to avoid race
conditions.  The immediate benefit is that all logic is now inside
`libpod`; the code isn't scattered anymore.

Further note that Docker does not support copying to tmpfs mounts.

Tests have been extended to cover *both* path resolutions for running
and created containers.  New tests have been added to exercise the
tmpfs-mount case.

Signed-off-by: Valentin Rothberg <rothberg@redhat.com>

@containers/podman-maintainers PTAL
@edsantiago, I'd love your input especially on the test changes

### CODE REVIEW

**Comment by vrothberg** [comment]
@TomSweeneyRedHat FWIW, I would _love_ to get this into 3.0 for two reasons: 1) It's a really thing and customers are asking for it (we can consider it a bug), 2) it will make backports much easier (there are still some bugs we need to iron out in cp).

Is a backport for RHEL possible?

File: libpod/container_copy_linux.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	decompressed, err := archive.DecompressStream(reader)
+	if err != nil {
+		c.unmount(false)
+		return nil, err
+	}
+
+	idMappings, idPair, err := getIDMappingsAndPair(c, mountPoint)
+	if err != nil {
+		c.unmount(false)
+		return nil, err
```
Comment (rhatdan) [inline libpod/container_copy_linux.go:60]: Should this have `decompressed.Close()`

**Review by rhatdan — APPROVED** [review]
Nice work.

File: libpod/container_copy_linux.go (hunk 1 above)
Comment (vrothberg) [reply libpod/container_copy_linux.go:60]: Good catch!  Yes, we need to close in this return path as well.

File: libpod/container_copy_linux.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+func (c *Container) copyFromArchive(ctx context.Context, path string, reader io.Reader) (func() error, error) {
+	// Make sure that "/" copies the *contents* of the mount point and not
+	// the directory.
+	if path == "/" {
+		path = "/."
+	}
+
+	// NOTE: make sure to unmount in error paths.
+	mountPoint, err := c.mount()
```
Comment (mheon) [inline libpod/container_copy_linux.go:34]: Probably should be conditional on the container not being mounted (c.state.Mounted)

Comment (vrothberg) [reply libpod/container_copy_linux.go:34]: Nice idea!  That should speed things up.  Added that to all places.

**Comment by edsantiago** [comment]
Just an FYI: in one of my test runs, I encountered a flake:
```
 ✗ podman cp file from host to container
...
$ podman start -a cpcontainer
   random-0-qvvUDTQtME
   Error: unable to start container f0191a2a6a86a3e78252b8d095a50a24b21980ea80f8c71de8df9c98703b9428: error attaching to container f0191a2a6a86a3e78252b8d095a50a24b21980ea80f8c71de8df9c98703b9428: read unixpacket @->/proc/self/fd/13/attach: read: connection reset by peer
   [ rc=125 (** EXPECTED 0 **) ]
```
"read unixpacket" has a horrible history of flakiness, usually relating to conmon: #7228, #3302. This was on my laptop, f33, conmon-2.0.26-1.fc33, 5.10.19-200.fc33.

**Comment by edsantiago** [comment]
May I make a friendly suggestion? This shaves over one minute from the test's run time (126 seconds down to 58):
[cp-fast.diff.txt](https://github.com/containers/podman/files/6076975/cp-fast.diff.txt)

Reason: `podman rm -f` has to wait those stupid ten seconds, and there's no `--do-it-right-now-dammit` option.

File: test/system/065-cp.bats [hunk 3]
```diff
@@ -68,6 +72,38 @@ load helpers
...
+    srcdir=$PODMAN_TMPDIR/cp-test-file-host-to-ctr
+    mkdir -p $srcdir
+    content=$(random_string 20)
+    echo $content > $srcdir/file
+
+    # RUNNING container
+    run_podman run -d --mount type=tmpfs,dst=/tmp --name cpcontainer $IMAGE sleep infinity
+    run_podman cp $srcdir/file cpcontainer:/tmp
+    run_podman exec cpcontainer cat /tmp/file
+    is "$output" "${content}" "cp to container's tmpfs"
```
Comment (edsantiago) [inline test/system/065-cp.bats:98]: A lot of people—myself first among those—are simple monkeys who, when we see an error message, grep directly for it in the test sources to find the relevant place. When error messages are identical in multiple places, brainless monkeys like me get confused.

Would you consider changing this to "cp to running container's tmpfs", and line 105 to "cp to created container's tmpfs"?

File: test/system/065-cp.bats [hunk 4]
```diff
@@ -68,6 +72,38 @@ load helpers
...
+        run_podman rm -f cpcontainer
+    done < <(parse_table "$tests")
+
+    run_podman rmi -f $cpimage
+}
+
+@test "podman cp file from host to container tmpfs mount" {
+    srcdir=$PODMAN_TMPDIR/cp-test-file-host-to-ctr
+    mkdir -p $srcdir
+    content=$(random_string 20)
```
Comment (edsantiago) [inline test/system/065-cp.bats:91]: I've developed a new habit when writing tests like this: adding placeholder markers to random content.
```bash
   content=tmpfile-content-$(random_string 20)
```
The main benefit is, when there's a test failure, the `tmpfile-content` text makes it very easy for a maintainer to see the error in better context: to understand what the expectation is and where it's set. Also importantly, when reading a long error log, it makes it easier to see which random string is which, and ignore the red herrings.

File: test/system/065-cp.bats [hunk 5]
```diff
@@ -110,8 +149,23 @@ load helpers
...
+    run_podman create --name cpcontainer --workdir=/srv $cpimage
+    while read id src dest dest_fullname description; do
+        # dest may be "''" for empty table cells
+        if [[ $dest == "''" ]];then
+            unset dest
+        fi
+        run_podman cp cpcontainer:$src "$srcdir$dest"
+        run cat $srcdir$dest_fullname
+        is "$output" "${randomcontent[$id]}" "$description (cp ctr:$src to \$srcdir$dest)"
+        rm $srcdir/$dest_fullname
```
Comment (edsantiago) [inline test/system/065-cp.bats:164]: Consistency: slash is not needed.

File: test/system/065-cp.bats [hunk 6]
```diff
@@ -110,8 +149,23 @@ load helpers
...
+    # Created container
+    run_podman create --name cpcontainer --workdir=/srv $cpimage
+    while read id src dest dest_fullname description; do
+        # dest may be "''" for empty table cells
+        if [[ $dest == "''" ]];then
+            unset dest
+        fi
+        run_podman cp cpcontainer:$src "$srcdir$dest"
+        run cat $srcdir$dest_fullname
+        is "$output" "${randomcontent[$id]}" "$description (cp ctr:$src to \$srcdir$dest)"
```
Comment (edsantiago) [inline test/system/065-cp.bats:163]: Hint:
```bash
    (remove the "run cat")
    is "$(< $srcdif$dest_fullname)" "${random...."
```
`$(< PATH)` is an idiomatic equivalent to `$(cat PATH)`

Comment (vrothberg) [reply test/system/065-cp.bats:163]: Nice!  I'll get that consistent across the copy tests.

File: test/system/065-cp.bats (hunk 4 above)
Comment (vrothberg) [reply test/system/065-cp.bats:91]: I'll add that. That can be super helpful!

File: test/system/065-cp.bats (hunk 3 above)
Comment (vrothberg) [reply test/system/065-cp.bats:98]: Sure!

**Comment by vrothberg** [comment]
> May I make a friendly suggestion? This shaves over one minute from the test's run time (126 seconds down to 58):
> [cp-fast.diff.txt](https://github.com/containers/podman/files/6076975/cp-fast.diff.txt)
> 
> Reason: `podman rm -f` has to wait those stupid ten seconds, and there's no `--do-it-right-now-dammit` option.

Yes, I was afraid of that.  I will add ` podman kill` before to cut things short :^)

**Comment by vrothberg** [comment]
NOTE: we currently cannot copy tty devices from a rootless container. One `e2e` test is disabled for now. I chatted with @nalind and he has an idea how to tackle that in the copier package.

I think we can merge. It's a very minor corner case and I will be adding a couple more commits to address other `cp` issues tomorrow.

**Comment by vrothberg** [comment]
Updated.  Thanks for the super helpful reviews!

File: libpod/container_api.go [hunk 7]
```diff
@@ -847,31 +843,57 @@ func (c *Container) ShouldRestart(ctx context.Context) bool {
...
-// ResolvePath resolves the specified path on the root for the container.  The
-// root must either be the mounted image of the container or the already
-// mounted container storage.
-//
-// It returns the resolved root and the resolved path.  Note that the path may
-// resolve to the container's mount point or to a volume or bind mount.
-func (c *Container) ResolvePath(ctx context.Context, root string, path string) (string, string, error) {
-	logrus.Debugf("Resolving path %q (root %q) on container %s", path, root, c.ID())
+// CopyFromArchive copies the contents of the specified tarStream to path.
+func (c *Container) CopyFromArchive(ctx context.Context, path string, tarStream io.Reader) (func() error, error) {
```
Comment (mheon) [inline libpod/container_api.go:847]: Can you make it more clear that this one copies into the container, the other one copies out of it? Updating the comments on both could help.

File: libpod/container_api.go [hunk 8]
```diff
@@ -847,31 +843,57 @@ func (c *Container) ShouldRestart(ctx context.Context) bool {
...
+
+		if err := c.syncContainer(); err != nil {
+			return nil, err
+		}
 	}
 
+	return c.copyToArchive(ctx, path, tarStream)
+}
+
+// Stat the specified path and return file info.
```
Comment (mheon) [inline libpod/container_api.go:874]: Can you make it clear that this stats the path *in the container*?

File: libpod/container_api.go [hunk 9]
```diff
@@ -847,31 +843,57 @@ func (c *Container) ShouldRestart(ctx context.Context) bool {
...
+func (c *Container) Stat(ctx context.Context, path string) (*define.FileInfo, error) {
 	if !c.batched {
 		c.lock.Lock()
 		defer c.lock.Unlock()
```
Comment (mheon) [inline libpod/container_api.go:878]: Should we consider making the container unlock after initial checks? Right now, `podman ps` while copying will block, which is not good - but at the same time, a container that stops while we are copying would be able to unmount the container from under us.

Comment (vrothberg) [reply libpod/container_api.go:878]: I think it's okay for `Stat` since it's fast, but I will look into optimizing it for the two copy cases.

**Comment by vrothberg** [comment]
@edsantiago 

> read: connection reset by peer

Just hit that in the F33 root system tests as well.

**Comment by vrothberg** [comment]
Repushing. Many jobs timed out (Cc @cevich).

**Comment by vrothberg** [comment]
> > May I make a friendly suggestion? This shaves over one minute from the test's run time (126 seconds down to 58):
> > [cp-fast.diff.txt](https://github.com/containers/podman/files/6076975/cp-fast.diff.txt)
> > Reason: `podman rm -f` has to wait those stupid ten seconds, and there's no `--do-it-right-now-dammit` option.
> 
> Yes, I was afraid of that. I will add ` podman kill` before to cut things short :^)

Unfortunately, `start -a`seems to be too flaky/broken. I'll change the copy tests but it's concerning.

**Comment by mheon** [comment]
We really ought to add `--timeout` to `podman rm`

**Comment by rhatdan** [comment]
I agree podman rm --timeout 0 -f is needed.

**Review by giuseppe — APPROVED** [review]
/lgtm
